### PR TITLE
Enable lint rules for Promise handling to discourage misuse of them

### PR DESCRIPTION
**Pull request by hughns**
Whilst working on the codebase I noticed that promises are not always being handled/treated correctly.

The purpose of enabling these lint rules to force a conscious decision around how Promises are handled including the errors.

The `@typescript-eslint/promise-function-async` rule is a stylistic one, so less important.

I'm anticipating this my prompt some discussion rather than just being merged as is 😇.

### CODE REVIEW

File: src/room/GroupCallView.tsx [hunk 1]
```diff
@@ -192,21 +184,31 @@ export const GroupCallView: FC<Props> = ({
...
     if (widget && preload && skipLobby) {
       // In preload mode without lobby we wait for a join action before entering
-      const onJoin = async (
-        ev: CustomEvent<IWidgetApiRequest>,
-      ): Promise<void> => {
-        defaultDeviceSetup(ev.detail.data as unknown as JoinCallData);
-        await enterRTCSession(rtcSession, perParticipantE2EE);
-        await widget!.api.transport.reply(ev.detail, {});
+      const onJoin = (ev: CustomEvent<IWidgetApiRequest>): void => {
+        defaultDeviceSetup(ev.detail.data as unknown as JoinCallData)
```
Comment (hughns) [inline src/room/GroupCallView.tsx:188]: Was it intended that `defaultDeviceSetup()` actually be run in parallel? I've assumed it should be serial.

File: src/room/GroupCallView.tsx [hunk 2]
```diff
@@ -192,21 +184,31 @@ export const GroupCallView: FC<Props> = ({
...
       };
       widget.lazyActions.on(ElementWidgetActions.JoinCall, onJoin);
       return (): void => {
         widget!.lazyActions.off(ElementWidgetActions.JoinCall, onJoin);
       };
     } else if (widget && !preload && skipLobby) {
       // No lobby and no preload: we enter the rtc session right away
-      defaultDeviceSetup({ audioInput: null, videoInput: null });
-      enterRTCSession(rtcSession, perParticipantE2EE);
+      defaultDeviceSetup({ audioInput: null, videoInput: null })
```
Comment (hughns) [inline src/room/GroupCallView.tsx:204]: Same as above.

File: src/room/GroupCallView.tsx (hunk 1 above)
Comment (toger5) [reply src/room/GroupCallView.tsx:188]: This might be the cause for the duplicated device requests we are getting!

**Comment by toger5** [comment]
Note: We brought this up in the standup today.
 - The general acceptance was positive.
 - Concerned that this might result in lots of lint disable comments. (looking through the changes we saw that for now its definitely manageable)

File: src/analytics/PosthogSpanProcessor.ts [hunk 3]
```diff
@@ -157,7 +161,8 @@ export class PosthogSpanProcessor implements SpanProcessor {
   /**
    * Shutdown the processor.
    */
-  public shutdown(): Promise<void> {
-    return Promise.resolve();
+  // eslint-disable-next-line @typescript-eslint/require-await
+  public async shutdown(): Promise<void> {
+    return;
```
Comment (robintown) [inline src/analytics/PosthogSpanProcessor.ts:166]: I would prefer to remove the `eslint-disable` comment and return `Promise.resolve()` in a case like this. It seems more in the spirit of the require-await lint rule.

File: src/config/Config.ts [hunk 4]
```diff
@@ -30,17 +30,21 @@ export class Config {
...
+    Config.internalInstance.initPromise = new Promise<void>(
+      (resolve, reject) => {
+        downloadConfig("../config.json")
+          .then((config) => {
+            Config.internalInstance.config = { ...DEFAULT_CONFIG, ...config };
+            resolve();
+          })
+          .catch(reject);
+      },
+    );
```
Comment (robintown) [inline src/config/Config.ts:47]: Could this be simplified to `downloadConfig("../config.json").then((config) => { ... })`?

File: src/rtcSessionHelpers.ts [hunk 5]
```diff
@@ -36,6 +36,7 @@ export function makeActiveFocus(): LivekitFocusActive {
   };
 }
 
+// eslint-disable-next-line @typescript-eslint/require-await
```
Comment (robintown) [inline src/rtcSessionHelpers.ts:39]: I suggest `Promise.resolve(...)`-ing the return value here as well. That way, the thing that makes eslint happy is closer to the comment at the bottom of the function about how we want to await something.

**Review by robintown — APPROVED** [review]
Thanks, the improved safety and error logging will be very welcome. Just some nits:

**Comment by toger5** [comment]
This has not moved to: https://github.com/element-hq/element-call/pull/2475

**Comment by toger5** [comment]
Also robins review has been applied in the squashed +rebased PR version.

**Comment by toger5** [comment]
Closing in favor of: #2607

**Pull request closed without merging**